Frameworks/UIKit/NSTextContainer.mm

```diff
@@ -62,38 +78,121 @@ - (NSArray*)exclusionPaths {
+                break;
+            }
+        }
+        if (path.origin.x + path.size.width > proposed.origin.x) {
+            proposed.origin.x = (path.origin.x + path.size.width) + padding;
+            proposed.size.width = maxWidth - proposed.origin.x;
+        }
+    }
+
+    return proposed;
+}
+
+// Finds the first point at which the given rect intersects the given exclusionContext, moving by delta each iteration
+static CGFloat __GetXPositionIntersectingZone(CGRect rect, const CGContextRef exclusionContext, CGFloat delta, CGFloat maxWidth) {
+    while (0 <= rect.origin.x && rect.origin.x < maxWidth &&
+           !CGContextIsPointInPath(exclusionContext, true, rect.origin.x + delta, rect.origin.y) &&
+           !CGContextIsPointInPath(exclusionContext, true, rect.origin.x + delta, rect.origin.y + rect.size.height / 2.0) &&
+           !CGContextIsPointInPath(exclusionContext, true, rect.origin.x + delta, rect.origin.y + rect.size.height)) {
+        rect.origin.x += delta;
+    }
+
+    return rect.origin.x;
+}
+
 /**
  @Status Caveat
  @Notes writingDirection and atIndex parameters are ignored
- TODO::1123 This needs to handle multiple exclusion zones properly when creating remainingRect
-            Currently processing exclusion zones left-to-right fails to calculate properly
+ * TODO 1465: handle writing directions
 */
 
 - (CGRect)lineFragmentRectForProposedRect:(CGRect)proposed
                                   atIndex:(NSUInteger)idx
                          writingDirection:(NSWritingDirection)direction
                             remainingRect:(CGRect*)remainingRect {
+    if (remainingRect) {
+        *remainingRect = CGRectZero;
+    }
+
     CGRect totalRect = CGRectMake(0, 0, _size.width, _size.height);
     CGRect ret = CGRectIntersection(proposed, totalRect);
+    if ([self.exclusionPaths count] == 0L || ret.size.width <= 0) {
+        return ret;
+    }
+
+    CGFloat padding = self.lineFragmentPadding;
+    if (padding <= 0.0) {
+        // If padding is invalid we will go into an infinite loop, use a default
+        padding = c_defaultPadding;
+    }
 
-    if (_exclusionPaths != nil) {
-        for (UIBezierPath* path in(NSArray*)_exclusionPaths) {
-            CGRect shapeRect = _CGPathFitRect(path.CGPath, ret, _size, self.lineFragmentPadding);
-            ret = CGRectIntersection(shapeRect, ret);
+    std::vector<CGRect> exclusionRectsForHorizontal{};
+    for (size_t i = 0; i < _exclusionPathBoundingRects.size(); ++i) {
+        CGRect boundingRect = _exclusionPathBoundingRects[i];
+        // If our proposed area doesn't intersect the bounding box of the exclusion zone at all, no need to compare against it
+        if ((proposed.origin.y < boundingRect.origin.y + boundingRect.size.height) &&
+            (proposed.origin.y + proposed.size.height > boundingRect.origin.y) &&
+            (proposed.origin.x < boundingRect.origin.x + boundingRect.size.width)) {
```

Comment:
CGRectIntersectsRect?

---
In reply to: [90091646](https://github.com/Microsoft/WinObjC/pull/1398#discussion_r90091646) [](ancestors = 90091646)